File: Mcs-cs/Mcs/Models/MaximumCommonSubgraph.cs
using System.Collections.Generic;
using System.Linq;

namespace Mcs.Models
{
    public class MaximumCommonSubgraph
    {
        public readonly Graph G1;
        public readonly Graph G2;
        /// <summary>
        /// List of vertex mappings to maximum common subgraph. For every pair, V1 is
        /// form G1 and V2 is from G2. List is sorted by V1.
        /// </summary>
        public readonly List<Pair> M;
        public readonly List<List<int>> AdjacencyMatrix;
        public readonly List<int> BestGenome;

        public MaximumCommonSubgraph(Graph g1, Graph g2, List<Pair> m, List<int> bestGenome)
        {
            G1 = g1;
            G2 = g2;
            M = m.OrderBy(pair => pair.V1).ToList();
            BestGenome = bestGenome;

            // Build adjacency matrix
            int n = M.Count;
            AdjacencyMatrix = new List<List<int>>();
            for (int i = 0; i < n; i++)
            {
                AdjacencyMatrix.Add(new List<int>());
                for (int j = 0; j < n; j++)
                {
                    AdjacencyMatrix[i].Add(G1.AdjacencyMatrix[M[i].V1][M[j].V1]);
                }
            }
        }

        public MaximumCommonSubgraph(Graph g1, Graph g2, List<Pair> M) : this(g1, g2, M, null) { }

        /// <returns>Sorted list of vertices from G1 not in the maximum common subgraph</returns>
        public List<int> VerticesFromG1NotInMcs()
        {
            List<int> v1s = M.Select(pair => pair.V1).ToList();
            return Enumerable.Range(0, G1.N).Except(v1s).ToList();
        }

        /// <returns>Sorted list of vertices from G2 not in the maximum common subgraph</returns>
        public List<int> VerticesFromG2NotInMcs()
        {
            List<int> v2s = M.Select(pair => pair.V2).OrderBy(v2 => v2).ToList();
            return Enumerable.Range(0, G2.N).Except(v2s).ToList();
        }

        public bool FromGenetic()
        {
            return BestGenome != null;
        }
    }
}

File: mcs-java/src/main/java/anc/models/MaximumCommonSubgraph.java
package anc.models;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import com.google.common.base.Objects;

public class MaximumCommonSubgraph {
    public final Graph g1;
    public final Graph g2;
    /**
     * List of vertex mappings to maximum common subgraph. For every pair, v1 is
     * form g1 and v2 is from g2. List is sorted by v1.
     */
    public final List<Pair> M;
    public final int[][] adjacencyMatrix;
    public final List<Integer> bestGenome;

    public MaximumCommonSubgraph(List<Pair> M, Graph g1, Graph g2, List<Integer> bestGenome) {
        this.M = M;
        this.g1 = g1;
        this.g2 = g2;
        this.bestGenome = bestGenome;

        // Build adjacency matrix
        final int n = M.size();
        this.adjacencyMatrix = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                adjacencyMatrix[i][j] = g1.M[M.get(i).v1][M.get(j).v1];
            }
        }
    }

    public MaximumCommonSubgraph(List<Pair> M, Graph g1, Graph g2) {
        this(M, g1, g2, null);
    }

    /** Results are sorted indeces */
    public List<Integer> verticesFromG1NotInMcs() {
        final List<Integer> result = IntStream.range(0, g1.n).boxed().collect(Collectors.toList());
        final List<Integer> v1s = M.stream().map(p -> p.v1).collect(Collectors.toList());
        result.removeAll(v1s);
        return result;
    }

    /** Results are sorted indeces */
    public List<Integer> verticesFromG2NotInMcs() {
        final List<Integer> result = IntStream.range(0, g2.n).boxed().collect(Collectors.toList());
        final List<Integer> v2s = M.stream().map(p -> p.v2).sorted().collect(Collectors.toList());
        result.removeAll(v2s);
        return result;
    }

    public boolean fromGenetic() {
        return bestGenome != null;
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(g1, g2, M, adjacencyMatrix, bestGenome);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        MaximumCommonSubgraph other = (MaximumCommonSubgraph) obj;
        return Objects.equal(g1, other.g1) && Objects.equal(g2, other.g2) && Objects.equal(M, other.M)
                && Objects.equal(adjacencyMatrix, other.adjacencyMatrix) && Objects.equal(bestGenome, other.bestGenome);
    }

}
